proxy.py

```diff
@@ -2,225 +2,145 @@
-    errors = {}
-    for platform, ycm in sorted(ycms.ycms.iteritems(), reverse=True):
-        result = ycm.parse(data['file'], data['line'], data['ch'])
-        if result is None:
-            continue
-        for error in result:
-            error_key = (error['kind'], error['location']['line_num'], error['text'])
-            if error_key in errors:
-                errors[error_key]['platforms'].append(platform)
-            else:
-                error['platforms'] = [platform]
-                errors[error_key] = error
-
-    return jsonify(
-        errors=errors.values()
-    )
-
-
-@app.route('/ycm/<process_uuid>/goto', methods=['POST'])
-def go_to(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    ycms = mapping[process_uuid]
-    data = request.get_json(force=True)
-    if 'patches' in data:
-        ycms.filesync.apply_patches(data['patches'])
-    for platform, ycm in sorted(ycms.ycms.iteritems(), reverse=True):
-        ycm.parse(data['file'], data['line'], data['ch'])
-        result = ycm.go_to(data['file'], data['line'], data['ch'])
-        if result is not None:
-            return jsonify(location=result)
-    return jsonify(location=None)
-
-
-@app.route('/ycm/<process_uuid>/create', methods=['POST'])
-def create_file(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    ycms = mapping[process_uuid]
-    data = request.get_json(force=True)
-    ycms.filesync.create_file(data['filename'], data['content'])
-    return 'ok'
-
-
-@app.route('/ycm/<process_uuid>/delete', methods=['POST'])
-def delete_file(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    ycms = mapping[process_uuid]
-    data = request.get_json(force=True)
-    ycms.filesync.delete_file(data['filename'])
-    return 'ok'
-
-
-@app.route('/ycm/<process_uuid>/ping', methods=['POST'])
-def ping(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    for ycm in mapping[process_uuid].ycms.itervalues():
-        if not ycm.ping():
-            return 'failed', 504
-
-    return 'ok'
 
+                send(id, result)
+        except (websocket.WebSocketException, geventwebsocket.WebSocketError, TypeError) as e:
+            # WebSocket closed
+            alive[0] = False
+        except Exception as e:
+            alive[0] = False
+            raise
+
+    # Functions to send back a response to a message, with its message ID.
+    def send_response(id, response):
+        if not isinstance(response, collections.Mapping):
+            response = dict(message=response)
+        response['_ws_message_id'] = id
+        return server_ws.send(json.dumps(response))
+
+
+    def send_error(id, message):
+        response = (dict(success=False, error=message))
+        send_response(id, response)
+
+    # Spawn a Greenlet to deal with the WebSocket connection
+    group = gevent.pool.Group()
```

Comment:
Ah, I just adapted the greenlet bit from https://github.com/pebble/cloudpebble-qemu-controller/blob/826a26b6e65ad15ded9771109f4e1b5399e8f91e/controller.py#L127 ... 
But of course, you use a group there because there are two processes! I did have a look at this and wonder why I'd create a greenlet and then block on it... but then I left it like this :p

I'll just take all the code out of do_recv() and run it in the server_ws function directly, and not use a greenlet.